src/monitors4codegen/monitor_guided_decoding/openai_gen.py

```diff
@@ -0,0 +1,167 @@
+"""
+This module provides the functions and classes for running Monitor-Guided Decoding over OpenAI models
+"""
+
+from enum import Enum
+import time
+from typing import List, Set
+import torch
+import asyncio
+
+from openai import OpenAI
+from monitors4codegen.monitor_guided_decoding.monitor import Monitor
+from monitors4codegen.monitor_guided_decoding.tokenizer_wrapper import TikTokenWrapper
+
+class OpenAI_Models(Enum):
+    TD3 = 'text-davinci-003'
+
+def openai_mgd(
+    client: OpenAI,
+    model: OpenAI_Models,
+    tokenizer: TikTokenWrapper,
+    prompt_tokenized: torch.Tensor,
+    temp: float,
+    top_p: float,
+    monitor: Monitor,
+    num_new_tokens: int
+):
+    """
+    This function generates completions with OpenAI models using the Monitor-Guided Decoding scheme.
+    """
+    prompt_tokenized: torch.Tensor = torch.tensor(prompt_tokenized, dtype=torch.int64)
+    assert len(prompt_tokenized.shape) == 1
+
+    all_tokens: torch.Tensor = prompt_tokenized
+    gen_text: bytes = b''
+    
+    gen_tokens: List[int] = []
+
+    tokens_sort_key = {k:[0, 0] for k in tokenizer.all_token_ids}
+    
+    # # TODO: Find a way to prioritize tokens to be blacklisted
```

Comment:
Please document why this TODO can improve the current implementation.

Reply:
Added in the PR: https://github.com/microsoft/monitors4codegen/pull/16